releases/release-4.0.12.md

```diff
@@ -0,0 +1,117 @@
+---
+title: TiDB 4.0.12 Release Notes
+---
+
+# TiDB 4.0.12 Release Notes
+
+Release date: March 31, 2021
+
+TiDB version: 4.0.12
+
+## New Features
+
++ TiFlash
+
+    - Add tools to check exact status of `tiflash replica` for online rolling update
+
+## Improvements
+
++ TiDB
+
+    - Refine explain info for batch cop [#23164](https://github.com/pingcap/tidb/pull/23164)
+    - Add warning info for exprs that can not be pushed to storage layer for explain statement. [#23020](https://github.com/pingcap/tidb/pull/23020)
+    - Ddl: migrate part of ddl package code from Execute/ExecRestricted to safe API (2) [#22935](https://github.com/pingcap/tidb/pull/22935)
+    - Ddl: migrate part of ddl package code from Execute/ExecRestricted to safe API (1). [#22929](https://github.com/pingcap/tidb/pull/22929)
+    - Add optimization-time and wait-TS-time into the slow log [#22918](https://github.com/pingcap/tidb/pull/22918)
+    - Support query partition_id from infoschema.partitions [#22489](https://github.com/pingcap/tidb/pull/22489)
+    - Add 'last_plan_from_binding' to help know whether sql's plan is matched with the hints in the binding [#21430](https://github.com/pingcap/tidb/pull/21430)
+    - Do not report error for prepared stmt execution if tidb_snapshot is set [#22641](https://github.com/pingcap/tidb/pull/22641)
+    - Record prepare execute fail as "Failed Query OPM" in monitor [#22672](https://github.com/pingcap/tidb/pull/22672)
+    - Add three format specifier for str_to_date expression [#22812](https://github.com/pingcap/tidb/pull/22812)
+    - Scattering truncated tables without pre-split option. [#22872](https://github.com/pingcap/tidb/pull/22872)
+
```

Comment:
+ TiKV
    - Pd_client: prevent a large number of reconnections in a short time [#9879](https://github.com/tikv/tikv/pull/9879)
    - Optimize seek write for many tombstones [#9729](https://github.com/tikv/tikv/pull/9729)
    - Change the default value of `leader-transfer-max-log-lag` to 128 to increase the success rate of leader transfer [#9605](https://github.com/tikv/tikv/pull/9605)